framework/src/transfers/MultiAppGeneralFieldNearestNodeTransfer.C

```diff
@@ -58,97 +63,118 @@ MultiAppGeneralFieldNearestNodeTransfer::prepareEvaluationOfInterpValues(
-    std::vector<std::vector<Point>> & points,
-    std::vector<std::vector<Real>> & local_values)
+MultiAppGeneralFieldNearestNodeTransfer::buildKDTrees(const unsigned int var_index)
 {
-  kdtrees.resize(_from_problems.size());
-  points.resize(_from_problems.size());
-  local_values.resize(_from_problems.size());
-
-  // Construct a local mesh for each problem
-  for (unsigned int i_from = 0; i_from < _from_problems.size(); ++i_from)
+  unsigned int num_sources;
+  if (!_nearest_positions_obj)
+    num_sources = _from_problems.size();
+  else
+    num_sources = _nearest_positions_obj->getPositions(/*initial=*/false).size();
+
+  _local_kdtrees.resize(num_sources);
+  _local_points.resize(num_sources);
+  _local_values.resize(num_sources);
+  unsigned int max_leaf_size = 0;
+
+  // Construct a local KDTree for each source (subapp or positions)
+  for (unsigned int i_source = 0; i_source < num_sources; ++i_source)
   {
-    FEProblemBase & from_problem = *_from_problems[i_from];
-    auto & from_mesh = from_problem.mesh(_displaced_source_mesh);
-    MooseVariableFieldBase & from_var =
-        from_problem.getVariable(0,
-                                 _from_var_names[var_index],
-                                 Moose::VarKindType::VAR_ANY,
-                                 Moose::VarFieldType::VAR_FIELD_ANY);
-
-    System & from_sys = from_var.sys().system();
-    unsigned int from_var_num = from_sys.variable_number(getFromVarName(var_index));
-    // FEM type info
-    auto & fe_type = from_sys.variable_type(from_var.number());
-    bool is_nodal = fe_type.family == LAGRANGE;
-
-    if (is_nodal)
+    // Nest a loop on apps only for the Positions case. Regular case only looks at one app
+    unsigned int num_apps_to_consider = _nearest_positions_obj ? _from_problems.size() : 1;
+    for (unsigned int app_index = 0; app_index < num_apps_to_consider; app_index++)
     {
-      for (const auto & node : from_mesh.getMesh().local_node_ptr_range())
+      unsigned int i_from;
+      if (!_nearest_positions_obj)
+        i_from = i_source;
+      else
+        i_from = app_index;
+      // NOTE: If we decide we do not want to mix apps when using the nearest-positions
+      // we could continue here when the positions is not the nearest to the app
+
+      FEProblemBase & from_problem = *_from_problems[i_from];
+      auto & from_mesh = from_problem.mesh(_displaced_source_mesh);
+      MooseVariableFieldBase & from_var =
+          from_problem.getVariable(0,
+                                   _from_var_names[var_index],
+                                   Moose::VarKindType::VAR_ANY,
+                                   Moose::VarFieldType::VAR_FIELD_ANY);
+
+      System & from_sys = from_var.sys().system();
+      unsigned int from_var_num = from_sys.variable_number(getFromVarName(var_index));
+      // FEM type info
+      auto & fe_type = from_sys.variable_type(from_var.number());
+      bool is_nodal = fe_type.family == LAGRANGE;
```

Comment:
```suggestion
      const unsigned int from_var_num = from_sys.variable_number(getFromVarName(var_index));
      // FEM type info
      const auto & fe_type = from_sys.variable_type(from_var.number());
      const bool is_nodal = fe_type.family == LAGRANGE;
```

Reply:
@roystgnr's favorite, `is_nodal` :smile:

Reply:
cleaning this up in another PR